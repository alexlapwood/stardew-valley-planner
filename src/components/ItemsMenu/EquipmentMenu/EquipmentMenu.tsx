import * as React from "react";

import * as cn from "classnames";

import BigText from "../../BigText/BigText";
import Sprite from "../../Sprite/Sprite";

// tslint:disable-next-line:no-var-requires
const equipment: IEquipment = require("../../../data/sdv.json").equipment;

interface IProps {
  isVisible: boolean;
  selectEquipment: (equipmentId: string, skinIndex: number) => void;
  selectedItem?: ISelectedItem;
}

const EquipmentMenu: React.SFC<IProps> = props => {
  const { isVisible, selectEquipment, selectedItem } = props;

  return isVisible ? (
    <div className="EquipmentMenu sdv-list">
      {Object.keys(equipment).map((equipmentId, equipmentIndex) =>
        equipment[equipmentId].skins.map((skin, skinIndex) => {
          let sprite;

          switch (equipmentId) {
            case "scarecrow":
              sprite = (
                <Sprite
                  height={16}
                  src="images/equipment.png"
                  width={16}
                  x={equipmentIndex * 16}
                  y={16 + skinIndex * 32}
                />
              );
              break;
            case "fence":
              sprite = (
                <div style={{ zoom: 0.5 }}>
                  <Sprite
                    height={32}
                    src="images/fences.png"
                    width={16}
                    x={skinIndex * 48}
                    y={0}
                  />
                  <Sprite
                    height={32}
                    src="images/fences.png"
                    width={16}
                    x={skinIndex * 48 + 32}
                    y={0}
                  />
                </div>
              );
              break;
            case "flooring":
              sprite = (
                <div>
                  <Sprite
                    height={16}
                    src="images/flooring.png"
                    width={16}
                    x={(skinIndex % 4) * 64}
                    y={Math.floor(skinIndex / 4) * 64}
                  />
                </div>
              );
              break;
            default:
              sprite = (
                <Sprite
                  height={16}
                  src="images/equipment.png"
                  width={16}
                  x={equipmentIndex * 16}
                  y={32 + skinIndex * 32}
                />
              );
              break;
          }

          return (
            <div
              className={cn("sdv-list-item", {
                selected:
                  selectedItem !== undefined &&
                  selectedItem.type === "equipment" &&
                  selectedItem.id === equipmentId &&
                  selectedItem.skinIndex === skinIndex
              })}
              data-automationid={`equipment--${equipmentId}`}
              key={skin}
              // tslint:disable-next-line:jsx-no-lambda
              onClick={() => {
                selectEquipment(equipmentId, skinIndex);
              }}
            >
              <div className="sdv-list-item-icon">{sprite}</div>
              <div className="sdv-list-item-text">
                <BigText>{skin}</BigText>
              </div>
            </div>
          );
        })
      )}
    </div>
  ) : null;
};

export default EquipmentMenu;
