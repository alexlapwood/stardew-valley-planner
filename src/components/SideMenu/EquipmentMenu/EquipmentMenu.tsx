import cn from "clsx";
import { createSignal, For } from "solid-js";

import { equipmentIds } from "../../../data/sdv.json";
import EquipmentMenuItem from "./EquipmentMenuItem/EquipmentMenuItem";

interface IProps {
  date: number;
  isVisible: boolean;
  range?: { from?: number; to?: number };
  selectEquipment: (equipmentId: string, skinIndex: number) => void;
  selectedItem?: ISelectedItem;
}

export default function EquipmentMenu(props: IProps) {
  const [currentSkins, setCurrentSkins] = createSignal<{
    [index: string]: number;
  }>({});

  return (
    <div class={cn("EquipmentMenu", "sdv-list", !props.isVisible && "hidden")}>
      <For
        each={equipmentIds.filter((_, i) => {
          if (props.range !== undefined) {
            if (props.range.from !== undefined && i < props.range.from) {
              return false;
            }
            if (props.range.to !== undefined && i > props.range.to) {
              return false;
            }
          }
          return true;
        })}
      >
        {(equipmentId) => (
          <EquipmentMenuItem
            date={props.date}
            equipmentId={equipmentId}
            selectEquipment={(equipmentId: string, skinIndex: number) => {
              setCurrentSkins({
                ...currentSkins(),
                [equipmentId]: skinIndex,
              });

              props.selectEquipment(equipmentId, skinIndex);
            }}
            selectedItem={props.selectedItem}
          />
        )}
      </For>
    </div>
  );
}
