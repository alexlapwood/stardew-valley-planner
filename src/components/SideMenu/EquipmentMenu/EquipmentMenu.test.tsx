import { mount } from "enzyme";
import React from "react";

import EquipmentMenu from "./EquipmentMenu";

// tslint:disable-next-line:no-var-requires
const { equipmentIds } = require("../../../data/sdv.json") as {
  equipmentIds: string[];
};

describe("<EquipmentMenu />", () => {
  it("renders correctly", () => {
    const equipmentMenu = mount(
      <EquipmentMenu
        date={0}
        isVisible={true}
        range={{ to: 2 }}
        selectEquipment={jest.fn()}
      />
    );

    expect(equipmentMenu).toMatchSnapshot();
  });

  describe("speciying a range", () => {
    it("renders the items up to and including range.to", () => {
      const to = 4;

      const equipmentMenu = mount(
        <EquipmentMenu
          date={0}
          range={{ to }}
          isVisible={true}
          selectEquipment={jest.fn()}
        />
      );

      expect(equipmentMenu.find("MenuItem")).toHaveLength(to + 1);
    });

    it("only renders the items from range.from onwards", () => {
      const from = 4;

      const equipmentMenu = mount(
        <EquipmentMenu
          date={0}
          range={{ from }}
          isVisible={true}
          selectEquipment={jest.fn()}
        />
      );

      expect(equipmentMenu.find("MenuItem")).toHaveLength(
        equipmentIds.length - 4
      );
    });

    it("only renders the items in the range", () => {
      const from = 2;
      const to = 4;
      const equipmentMenu = mount(
        <EquipmentMenu
          date={0}
          range={{ from, to }}
          isVisible={true}
          selectEquipment={jest.fn()}
        />
      );

      expect(equipmentMenu.find("MenuItem")).toHaveLength(to - from + 1);
    });
  });
});
