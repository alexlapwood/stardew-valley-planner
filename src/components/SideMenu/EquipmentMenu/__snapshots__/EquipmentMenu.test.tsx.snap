// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<EquipmentMenu /> > renders correctly 1`] = `
<div>
  <div
    class="EquipmentMenu sdv-list"
  >
    <div>
      <div
        class="sdv-list-item flex-horizontal"
        data-testid="equipment-scarecrow--0"
      >
        <div
          class="sdv-list-item-icon"
        >
          <div
            style="height: 16px; transform: scale(0.5); transform-origin: top center; white-space: nowrap; width: 16px;"
          >
            <div
              class="sprite"
              style="display: inline-block; vertical-align: top; background-image: url(images/equipment-sheet.png); background-position: -0px -0px; height: 32px; width: 16px;"
            />
          </div>
        </div>
        <div
          class="sdv-list-item-text flex margin-right"
        >
          <div
            class="BigText"
          >
            <div
              style="transform-origin: left center; white-space: nowrap; transform: scale(1);"
            >
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -408px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -536px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -520px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -656px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -552px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -536px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -656px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -632px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -696px -0px; height: 16px; width: 8px;"
              />
            </div>
          </div>
        </div>
        <button
          class="margin-right sdv-button-down"
          data-testid="equipment-dropdown--trigger"
        />
      </div>
      
    </div>
    <div>
      <div
        class="sdv-list-item flex-horizontal"
        data-testid="equipment-sprinkler--0"
      >
        <div
          class="sdv-list-item-icon"
        >
          <div
            class="sprite"
            style="display: inline-block; vertical-align: top; background-image: url(images/equipment-sheet.png); background-position: -144px -16px; height: 16px; width: 16px;"
          />
        </div>
        <div
          class="sdv-list-item-text flex margin-right"
        >
          <div
            class="BigText"
          >
            <div
              style="transform-origin: left center; white-space: nowrap; transform: scale(1);"
            >
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -408px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -640px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -656px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -584px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -624px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -600px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -608px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -552px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -656px -0px; height: 16px; width: 8px;"
              />
            </div>
          </div>
        </div>
        <button
          class="margin-right sdv-button-down"
          data-testid="equipment-dropdown--trigger"
        />
      </div>
      
    </div>
    <div>
      <div
        class="sdv-list-item flex-horizontal"
        data-testid="equipment-fence--0"
      >
        <div
          class="sdv-list-item-icon"
        >
          <div
            style="height: 16px; transform: scale(0.5); transform-origin: top left; white-space: nowrap; width: 16px;"
          >
            <div
              class="sprite"
              style="display: inline-block; vertical-align: top; background-image: url(images/fences.png); background-position: -0px -0px; height: 32px; width: 16px;"
            />
            <div
              class="sprite"
              style="display: inline-block; vertical-align: top; background-image: url(images/fences.png); background-position: -32px -0px; height: 32px; width: 16px;"
            />
          </div>
        </div>
        <div
          class="sdv-list-item-text flex margin-right"
        >
          <div
            class="BigText"
          >
            <div
              style="transform-origin: left center; white-space: nowrap; transform: scale(1);"
            >
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -440px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -632px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -632px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -544px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -0px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -304px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -552px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -624px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -536px -0px; height: 16px; width: 8px;"
              />
              <div
                class="sprite"
                style="display: inline-block; vertical-align: top; background-image: url(images/sdv-font.png); background-position: -552px -0px; height: 16px; width: 8px;"
              />
            </div>
          </div>
        </div>
        <button
          class="margin-right sdv-button-down"
          data-testid="equipment-dropdown--trigger"
        />
      </div>
      
    </div>
  </div>
</div>
`;
